File: LeetCode/Medium/2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.py
class SmallestInfiniteSet:

    def __init__(self):
        self.heap = []
        self.curr = 1

    def popSmallest(self) -> int:
        if self.heap:
            return self.heap.pop()
        else:
            self.curr += 1
            return self.curr - 1

    def addBack(self, num: int) -> None:
        if num < self.curr and num not in self.heap:
            self.heap.append(num)
            self.heap.sort(reverse=True)


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)

File: LeetCode/Medium/2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.cpp
class SmallestInfiniteSet {
public:
    bool heap[1001]{false,};
    int curr = 1;
    
    SmallestInfiniteSet() {
        
    }
    
    int popSmallest() {
        int ret = 0;
        for (int i = curr; i < 1001; ++i){
            if (!heap[i]){
                heap[i] = true;
                ret = i;
                break;
            }
        }
        curr = ret + 1;
        return ret;
    }
    
    void addBack(int num) {
        if (heap[num]){
            if (num < curr) curr = num;
            heap[num] = false;
        }
    }
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */